packages/babel-plugin-transform-typescript/src/namespace.js

```diff
@@ -0,0 +1,177 @@
+      case "FunctionDeclaration":
+      case "ClassDeclaration":
+        names[subNode.id.name] = true;
+        continue;
+      case "VariableDeclaration":
+        for (const variable of subNode.declarations) {
+          names[variable.id.name] = true;
+        }
+        continue;
+      default:
+        // Neither named declaration nor export, continue to next item.
+        continue;
+      case "ExportNamedDeclaration":
+      // Export declarations get parsed using the next switch.
+    }
+
+    // Transform the export declarations that occur inside of a namespace.
+    switch (subNode.declaration.type) {
+      case "TSEnumDeclaration":
+      case "FunctionDeclaration":
+      case "ClassDeclaration": {
+        const itemName = subNode.declaration.id.name;
+        names[itemName] = true;
+        namespaceTopLevel.splice(
+          i++,
+          1,
+          subNode.declaration,
+          t.expressionStatement(
+            t.assignmentExpression(
+              "=",
+              getMemberExpression(t, name, itemName),
+              t.identifier(itemName),
+            ),
+          ),
+        );
+        break;
+      }
+      case "VariableDeclaration":
+        if (subNode.declaration.kind !== "const") {
+          throw path.hub.file.buildCodeFrameError(
+            subNode.declaration,
+            "Namespaces exporting non-const are unsupported.",
+          );
+        }
+        for (const variable of subNode.declaration.declarations) {
+          variable.init = t.assignmentExpression(
+            "=",
+            getMemberExpression(t, name, variable.id.name),
+            variable.init,
+          );
+        }
+        namespaceTopLevel[i] = subNode.declaration;
+        break;
+      case "TSModuleDeclaration": {
+        const transformed = handleNested(path, t, subNode.declaration, name);
+        const moduleName = subNode.declaration.id.name;
+        if (names[moduleName]) {
+          namespaceTopLevel[i] = transformed;
+        } else {
+          names[moduleName] = true;
+          namespaceTopLevel.splice(
+            i++,
+            1,
+            getDeclaration(t, moduleName),
+            transformed,
+          );
+        }
+      }
+    }
+  }
+
+  // {}
+  let fallthroughValue = t.objectExpression([]);
+
+  if (parentExportName) {
+    // _A.B = {}
+    const exportAssignment = t.assignmentExpression(
+      "=",
+      getMemberExpression(t, parentExportName, realName),
+      fallthroughValue,
+    );
+    // _A.B || (_A.B = {})
+    fallthroughValue = t.logicalExpression(
+      "||",
+      getMemberExpression(t, parentExportName, realName),
+      exportAssignment,
+    );
+  }
+
+  // B = ...
+  const assignment = t.assignmentExpression(
+    "=",
+    t.identifier(realName),
+    fallthroughValue,
+  );
+  // B || (B = ...)
+  const parameter = t.logicalExpression(
+    "||",
+    t.identifier(realName),
+    assignment,
+  );
+
+  return t.expressionStatement(
```

Comment:
And here
```js
return template.statement.ast`
  (function (${name}) {
    ${namespaceTopLevel}
  })(${id} || (${id} = ${fallthroughValue}));
`;
```

Reply:
bb95958dd0c5fbf0ff41293dbb10dd1097d6e783